provider/datagen/src/driver.rs

```diff
@@ -954,84 +1012,118 @@ fn select_locales_for_key(
+                            // Root locale: do not include descendants (use `full` for that)
+                            Some((LanguageIdentifier::UND, LocaleFamilyAnnotations::single()))
+                        } else {
+                            // All other locales: copy the requested annotations
+                            Some((langid.clone(), *annotations))
+                        }
+                    } else {
+                        // Full locale family: set the bit instead of adding to the set
+                        debug_assert_eq!(annotations, &LocaleFamily::full().annotations);
+                        include_full = true;
+                        None
+                    }
+                })
+                .collect(),
+            LocalesWithOrWithoutFallback::WithoutFallback { langids } => langids
+                .iter()
+                // Map langids without fallback to the `single` family
+                .map(|langid| (langid.clone(), LocaleFamilyAnnotations::single()))
+                .collect(),
+        };
 
-    let mut it = locale_families.iter().filter_map(|x| {
-        x.langid
-            .as_ref()
-            .map(|langid| (langid, x.include_ancestors))
-    });
-    let ExplicitImplicitLocaleSets { explicit, implicit } =
-        make_explicit_implicit_sets(key, &mut it, &supported_map, fallbacker)?;
-
-    let supported_and_explicit = supported_map
-        .into_values()
-        .flatten()
-        .chain(explicit.iter().cloned());
-
-    // Need to check this ahead of time because we can avoid loading the fallbacker.
-    if locale_families.contains(&LocaleFamily::full()) {
-        return Ok(supported_and_explicit.collect());
+    if include_full && requested_families.is_empty() {
+        // Special case: return now so we don't need the fallbacker (and its requisite CLDR data)
+        let selected_locales = supported_map.into_values().flatten().collect();
+        return Ok(selected_locales);
     }
 
+    // Need the fallbacker now.
     let fallbacker = fallbacker.as_ref().map_err(|e| *e)?;
     let fallbacker_with_config = fallbacker.for_config(key.fallback_config());
 
-    // TODO(#4629): Allow the exclusion of und
-    let include_und = true;
-
-    let result = supported_and_explicit
-        .filter(|locale_orig| {
-            let mut locale = locale_orig.clone();
-            locale.remove_aux();
-            if implicit.contains(&locale) {
-                return true;
-            }
-            if explicit.contains(&locale) {
-                return true;
+    // Compute a map from LanguageIdentifiers to DataLocales, including inherited auxiliary keys
+    // and extensions. Also resolve the ancestors and descendants while building this map.
+    let mut selected_langids = requested_families.keys().cloned().collect::<HashSet<_>>();
+    let expansion_map: HashMap<&LanguageIdentifier, HashSet<DataLocale>> = supported_map
+        .keys()
+        .chain(requested_families.keys())
```

Comment:
does this duplicate work in the intersection?

Reply:
Yes, it duplicates work. Probably worthwhile avoiding the extra work. Fixed by collecting the langids into a set first.